luarules/gadgets/unit_objectify.lua

```diff
@@ -151,20 +187,81 @@ if gadgetHandler:IsSyncedCode() then
 else -- UNSYNCED
 
 
-    local CMD_MOVE = CMD.MOVE
-    local spGetUnitDefID = Spring.GetUnitDefID
+	local myAllyTeam = Spring.GetMyAllyTeamID()
+	local spectating = Spring.GetSpectatingState()
+	function gadget:PlayerChanged(playerID)
+		myAllyTeam = Spring.GetMyAllyTeamID()
+		spectating = Spring.GetSpectatingState()
+	end
+
+	-- "predicate" tables are checked in their index order
+	-- with early returns when the first check is matched:
+	local allyBeingBuilt = {
+		{ check = canRepair, command = CMD_REPAIR }, -- so this is the priority
+		{ check = canMove,   command = CMD_MOVE }, -- and this is the fallback
+	}
+	local allyObjectUnit = {
+		{ check = canReclaim, command = CMD_RECLAIM },
+		{ check = canMove,    command = CMD_MOVE },
+	}
+	local hideEnemyDecoy = {
+		{ check = canAttack,  command = CMD_ATTACK },
+		{ check = canReclaim, command = CMD_RECLAIM },
+		{ check = canMove,    command = CMD_MOVE },
+	}
 
-    function gadget:DefaultCommand(type, id, cmd)
-		if type == "unit" and cmd ~= CMD_MOVE then
-			local uDefID = spGetUnitDefID(id)
-			if isObject[uDefID] or isDecoration[uDefID] then
-				-- make sure a command given on top of a objectified/decoration unit is a move command
-				if select(4, Spring.GetUnitHealth(id)) == 1 then
-					return CMD_MOVE
+	local function scanSelection(predicates, limit)
```

Comment:
perhaps add `limit = limit or 64` so one doesn't need to pass a limit explicitly?